internal/tofu/context_plan.go

```diff
@@ -568,32 +563,100 @@ func (c *Context) postPlanValidateImports(importResolver *ImportResolver, allIns
-				Config: ic,
-			})
-		}
+		importTargets = append(importTargets, &ImportTarget{
+			Config: ic,
+		})
 	}
 	return importTargets
 }
 
-func (c *Context) validateImportTargets(config *configs.Config, importTargets []*ImportTarget) (diags tfdiags.Diagnostics) {
+// validateImportTargets makes sure all import targets are not breaking the following rules:
+//  1. Imports are attempted to resources that do not exist (if config generation is not enabled).
+//  2. Config generation is not attempted for resources inside sub-modules
+//  3. Config generation is not attempted for resources with indexes (for_each/count) - This will always include
+//     resources for which we could not yet resolve the address
+func (c *Context) validateImportTargets(config *configs.Config, importTargets []*ImportTarget, generateConfigPath string) (diags tfdiags.Diagnostics) {
+	configGeneration := len(generateConfigPath) > 0
 	for _, imp := range importTargets {
 		staticAddress := imp.StaticAddr()
 		descendantConfig := config.Descendent(staticAddress.Module)
+
+		// If import target's module does not exist
 		if descendantConfig == nil {
-			diags = diags.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Cannot import to non-existent resource address",
-				Detail:   fmt.Sprintf("Importing to resource address '%s' is not possible, because that address does not exist in configuration. Please ensure that the resource key is correct, or remove this import block.", staticAddress),
-				Subject:  imp.Config.DeclRange.Ptr(),
-			})
-			return
+			if configGeneration {
+				// Attempted config generation for resource in non-existing module. So error because resource generation
+				// is not allowed in a sub-module
+				diags = diags.Append(importConfigGenerationInModuleDiags(staticAddress.String(), imp.Config))
+			} else {
+				diags = diags.Append(importResourceWithoutConfigDiags(staticAddress.String(), imp.Config))
+			}
+			continue
+		}
+
+		if _, exists := descendantConfig.Module.ManagedResources[staticAddress.Resource.String()]; !exists {
+			if configGeneration {
+				if imp.ResolvedAddr() == nil {
+					// If we could not resolve the address of the import target, the address must have contained indexes
+					diags = diags.Append(importConfigGenerationWithIndexDiags(staticAddress.String(), imp.Config))
+					continue
+				} else if !imp.ResolvedAddr().Module.IsRoot() {
+					diags = diags.Append(importConfigGenerationInModuleDiags(imp.ResolvedAddr().String(), imp.Config))
+					continue
+				} else if imp.ResolvedAddr().Resource.Key != addrs.NoKey {
+					diags = diags.Append(importConfigGenerationWithIndexDiags(imp.ResolvedAddr().String(), imp.Config))
+					continue
+				}
+			} else {
+				diags = diags.Append(importResourceWithoutConfigDiags(staticAddress.String(), imp.Config))
+				continue
+			}
 		}
 	}
 	return
 }
 
+func importConfigGenerationInModuleDiags(addressStr string, config *configs.Import) *hcl.Diagnostic {
```

Comment:
I like that we pulled out the diag creations here, it makes it a bit more readable 👍